clippy_lints/src/matches.rs

```diff
@@ -892,6 +925,30 @@ fn check_match_single_binding(cx: &LateContext<'_, '_>, ex: &Expr<'_>, arms: &[A
     }
 }
 
+fn check_match_refcell_borrow(cx: &LateContext<'_, '_>, expr: &Expr) {
+    match &expr.kind {
+        ExprKind::MethodCall(_, _, args) => {
+            if let [arg] = &**args {
+                let method_def_id = cx.tables.type_dependent_def_id(expr.hir_id).unwrap();
+                if crate::utils::match_def_path(cx, method_def_id, &paths::REFCELL_BORROW) || crates::utils::match_def_path(cx, method_def_id, &paths::REFCELL_BORROWMUT) {
+                    span_lint_and_help(
+                        cx,
+                        REFCELL_BORROW,
+                        expr.span,
+                        "consider move this into a `let` binding",
+                    );
+                } else {
+                    check_match_refcell_borrow(cx, arg);
```

Comment:
You don't need to recursively walk expressions, since every expression is visited by `check_expr` anyway.